src/common/TermNames.h

```diff
@@ -38,49 +43,138 @@ class TermNames {
-    std::size_t size() const noexcept { return names.size(); }
+    TermName const * tryGetNameForTerm(PTRef term) const {
+        if (auto vecPtr = tryGetNamesForTerm(term)) { return &pickName(*vecPtr); }
+
+        return nullptr;
+    }
+
+    // Iterates over const names and their corresponding terms (according to value_type)
+    inline auto begin() const noexcept;
+    inline auto end() const noexcept;
+
+    bool empty() const noexcept { return scopedNames.empty(); }
+    std::size_t size() const noexcept { return scopedNames.size(); }
+
+    // A const view to just the names
+    inline auto names() const;
+    // A const view to just the terms
+    auto const & terms() const { return _terms; }
 
 protected:
     friend class MainSolver;
 
     using NameToTermMap = std::unordered_map<TermName, PTRef>;
     using TermToNamesMap = std::unordered_map<PTRef, std::vector<TermName>, PTRefHash>;
 
+    class Iterator {
+    public:
+        using NameIterator = ScopedVector<TermName>::const_iterator;
+        using TermIterator = std::vector<PTRef>::const_iterator;
+
+        explicit Iterator(NameIterator nit, TermIterator tit) : nameIterator{nit}, termIterator{tit} {}
+
+        value_type operator*() const { return {*nameIterator, *termIterator}; }
+
+        Iterator & operator++() {
+            ++nameIterator;
+            ++termIterator;
+            return *this;
+        }
+        Iterator operator++(int) {
+            auto it = *this;
+            operator++();
+            return it;
+        }
+
+        bool operator==(Iterator const & rhs) const noexcept { return nameIterator == rhs.nameIterator; }
+
+    protected:
+        NameIterator nameIterator;
+        TermIterator termIterator;
+    };
+
+    class NamesView;
+
     void pushScope() {
         if (isGlobal()) { return; }
-        names.pushScope();
+        scopedNames.pushScope();
     }
 
     void popScope() {
         if (isGlobal()) { return; }
-        names.popScope([this](TermName const & name) {
+        assert(scopedNames.size() == _terms.size());
+        scopedNames.popScope([this](TermName const & name) {
             auto it = nameToTerm.find(name);
             if (it == nameToTerm.end()) { return; }
             PTRef term = it->second;
-            assert(termToNames.find(term) != termToNames.end());
-            auto & names = termToNames.at(term);
-            names.erase(std::find(names.begin(), names.end(), name));
+            auto & names_ = const_cast<std::vector<TermName> &>(namesForTerm(term));
+            names_.erase(std::find(names_.begin(), names_.end(), name));
             nameToTerm.erase(it);
         });
+        assert(scopedNames.size() <= _terms.size());
+        _terms.resize(scopedNames.size());
     }
 
     SMTConfig const & config;
 
-    ScopedVector<TermName> names;
+    ScopedVector<TermName> scopedNames;
     NameToTermMap nameToTerm;
     TermToNamesMap termToNames;
+
+    std::vector<PTRef> _terms;
```

Comment:
Couldn't we have just a `ScopedVector<pair<TermName, PTRef>>`? Wouldn't it make things simpler?

Reply:
It makes it easier for the first commit, not easier for the rest commits. However, the first commit is more important, so I will use it this way.